File: Walktrough_UnitTesting-ObjectiveC/Walktrough_UnitTesting-ObjectiveC/CalculatorBrain.m
//
//  CalculatorBrain.m
//  Walktrough_UnitTesting-ObjectiveC
//
//  Created by MCS Devices on 11/2/17.
//  Copyright © 2017 Mobile Consulting Solutions. All rights reserved.
//

#import "CalculatorBrain.h"

@interface CalculatorBrain()

@property (nonatomic, strong) NSMutableArray * operandStack;

@end

@implementation CalculatorBrain

@synthesize operandStack = _operandStack;

- (NSMutableArray*) operandStack
{
    if(_operandStack==nil)
        _operandStack=[[NSMutableArray alloc]init];
    
    return _operandStack;
}

-(void) setOperandStack:(NSMutableArray *)operandStack
{
    _operandStack=operandStack;
}

- (void) pushOperand:(double) operand
{
    [self.operandStack addObject:[NSNumber numberWithDouble:operand]];
}

- (double)popOperand
{
    NSNumber*operandObject = [self.operandStack lastObject];
    
    if(operandObject!=nil)
        [self.operandStack removeLastObject];
    
    return [operandObject doubleValue];
}

-(double)performOperation:(NSString*)operation{
    double result =0;
    
    if ([operation isEqualToString:@"+"])
    {
        result=[self popOperand]+[self popOperand];
    }
    else if([operation isEqualToString:@"*"])
    {
        result=[self popOperand]*[self popOperand];
    }
    else if([operation isEqualToString:@"-"])
    {
        double operand2= [self popOperand];
        result =[self popOperand]-operand2;
    }
    else if([operation isEqualToString:@"/"])
    {
        double operand2= [self popOperand];
        result =[self popOperand]/operand2;
    }
    
    [self pushOperand:result];
    return result;
}

@end

File: Walktrough-UnitTesting-Swift/Walktrough-UnitTesting-Swift/CalculatorBrain.swift
//
//  CalculatorBrain.swift
//  Walktrough-UnitTesting-Swift
//
//  Created by MCS Devices on 11/2/17.
//  Copyright © 2017 Mobile Consulting Solutions. All rights reserved.
//

import UIKit

class CalculatorBrain: NSObject {

    
    var operandStack: NSMutableArray? = NSMutableArray()
    
    func setOperandStack (operandStack: NSMutableArray)
    {
        self.operandStack=operandStack
    }
    
    func pushOperand(operand: Double)
    {
        self.operandStack?.add(operand)
    }
    
    func popOperand() -> Double
    {
        if let operandObject = self.operandStack?.lastObject{
            
            if(operandObject != nil){
                self.operandStack?.removeLastObject()
            }
            return operandObject as! Double
            
        } else {
            return 0
        }
       
    }
    
    func performOperation(operation: String) -> Double {
    
        var result: Double = 0;
        
        if operation == "+"
        {
            result = popOperand() + popOperand()
        }
        else if operation == "*"
        {
            result = popOperand() * popOperand()
        }
        else if operation == "-"
        {
            let operand2 = popOperand()
            result = popOperand() - operand2 ;
        }
        else if operation == "/"
        {
            let operand2 = popOperand()
            result = popOperand()/operand2;
        }
        self.pushOperand(operand: result)
        return result;
    }

    
}
